### PR TITLE
API: MultiIndex.names|codes|levels returns tuples

### LINKED ISSUE #53531: API: Change MultiIndex.levels/codes to use `tuple` instead of `FrozenList`?
The doc of FrozenList says

```
    Container that doesn't allow setting item *but*
    because it's technically hashable, will be used
    for lookups, appropriately, etc.
```

which a `tuple` satisfies. Not sure why a tuple wasn't use originally but it may be simpler to use them instead of a custom `FrozenList` that acts like a tuple.

### PR DESCRIPTION
**Pull request by mroeschke**
- [ ] closes #53531 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by WillAyd** [comment]
Just playing devil's advocate but why would we prefer tuple here over `Index`? The latter is what we return from DataFrame.columns which maybe should harmonize with this

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -2925,7 +2931,9 @@ def _partial_tup_index(self, tup: tuple, side: Literal["left", "right"] = "left"
             if lab not in lev and not isna(lab):
                 # short circuit
                 try:
-                    loc = algos.searchsorted(lev, lab, side=side)
+                    # Argument 1 to "searchsorted" has incompatible type "Index";
```
Comment (WillAyd) [inline pandas/core/indexes/multi.py:2934]: This is surprising to see?

**Review by WillAyd — APPROVED** [review]
in spite of my comments this lgtm

**Comment by mroeschke** [comment]
> Just playing devil's advocate but why would we prefer tuple here over Index?

I picked tuple to align with the prior `FrozenList` which seemed to act like an immutable `list`. If we were to go with `Index` I think each would have to be `Index[object]

* `MultiIndex.codes` would be an Index of ndarrays
* `MultiIndex.levels` would be an Index of Indexes
* `MultiIndex.names` would be an Index of object/inferred dtype which would be reasonable.

But yeah the biggest thing is that the returned object should be able to be resized.

File: doc/source/user_guide/groupby.rst [hunk 2]
```diff
@@ -143,7 +143,7 @@ the columns except the one we specify:
 .. ipython:: python
 
    df2 = df.set_index(["A", "B"])
-   grouped = df2.groupby(level=df2.index.names.difference(["B"]))
+   grouped = df2.groupby(level="A")
```
Comment (rhshadrach) [inline doc/source/user_guide/groupby.rst:146]: The line leading into this is:

> If we also have a MultiIndex on columns ``A`` and ``B``, we can group by all
the columns except the one we specify:

If we want to preserve this, one can do `[e for e in df2.index.names if e not in {"A"}]`. But I'd also be okay to remove the entire example.

Comment (rhshadrach) [reply doc/source/user_guide/groupby.rst:146]: I'm worried taking union and difference here might be a common use case that we're breaking. I'm okay with removing the feature, but we could deprecate uses of these methods to make the breaking change a bit more soft. Of course, with our deprecation policy that would mean punting this off for some time.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: doc/source/user_guide/groupby.rst (hunk 2 above)
Comment (mroeschke) [reply doc/source/user_guide/groupby.rst:146]: Fair point. If you feel strongly about it we can go through a deprecation, but the fact that FrozenList was never really publically documented (except indirectly here with this example) makes me OK with breaking this use case.

As a data point, in cudf uses `FrozenList` in `cudf.MultiIndex` to match return types but has no usage of `FrozenList.union/difference`

Comment (rhshadrach) [reply doc/source/user_guide/groupby.rst:146]: I do not feel strongly here. Okay to go forward in my opinion.

**Comment by mroeschke** [comment]
Will merge this week unless there's any objections

File: pandas/core/indexes/multi.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/indexes/multi.py:2935]: Yeah, I think our typing isn't entirely correct somewhere

**Comment by rhshadrach** [comment]
@mroeschke - just the request to fix the docs: https://github.com/pandas-dev/pandas/pull/57042#discussion_r1472031470

**Comment by mroeschke** [comment]
> @mroeschke - just the request to fix the docs: [#57042 (comment)](https://github.com/pandas-dev/pandas/pull/57042#discussion_r1472031470)

Ah thanks for the reminder. Removed

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by jorisvandenbossche** [comment]
This is quite a breaking change .. I know you knew this and therefore kept it for 3.0, but I still do wonder if this is worth the breakage, and if we really want to change it, whether we couldn't first deprecate some aspects of it.

The breaking change I ran into with geopandas, is the fact that a tuple cannot be concatenated like a list. We actually used the same pattern in pandas itself, as the diff in this PR has this change in the groupby code:

```diff
-        mi = MultiIndex(levels=levels, codes=codes, names=idx.names + [None])
+        mi = MultiIndex(levels=levels, codes=codes, names=list(idx.names) + [None])
```

I can imagine quite some external libraries use this when manipulating index / multi-index objects. 

We could first deprecate those methods on FrozenList that won't work anymore for tuple (I see @rhshadrach mentioned the same on the issue for the `difference`/`union` methods)

**Pull request merged**